[Android] Enable more code and tests

finagolfin opened the pull request:
while disabling setting extended file attributes and a test creating a hard link, features not normally allowed on Android.

This got all the tests passing natively on Android 13 AArch64.

@hyp and @compnerd, try it out yourself and please review.

finagolfin commented on Sources/FoundationEssentials/TimeZone/TimeZone_Cache.swift:
```diff
@@ -173,7 +173,7 @@ struct TimeZoneCache : Sendable {
             }
             
-#if os(Linux) && !os(WASI)
+#if os(Linux)
```
@kateinoigakukun, I assume the `os` can't be both `Linux` and `WASI`?

kateinoigakukun replied:
Right, just a `os(Linux)` should be enough.

jmschonfeld commented on Sources/FoundationEssentials/FileManager/FileManager+Files.swift:
```diff
@@ -956,6 +956,8 @@ extension _FileManagerImpl {
                 // WASI does not support extended attributes
                 throw CocoaError.errorWithFilePath(.featureUnsupported, path)
+                #elseif canImport(Android)
+                // Android doesn't allow setting this for normal apps, so just skip it.
```
If Android doesn't allow setting this, should we throw an error similarly to WASI, either `.featureUnsupported` or perhaps an error indicating that the operation isn't allowed?

finagolfin replied:
I considered simply enabling the WASI error block above on Android too, and would have if this method were just setting these extended attributes alone, but this `setAttributes()` method is also setting a bunch of other attributes and works fine for that on Android.

I don't know the situation on WASI but pretty much all files on Android appear to have some preset extended attributes, but normal users can't change them. Given that, it makes sense to just skip setting them on Android, whereas maybe WASI normally doesn't have them or something, so the error above is rare on that platform.

finagolfin commented:
Updated to allow reading from zero-sized system files on Android too and imported the Android overlay in a test file that now needs it and rebased, tested with the Sep. 4 trunk source snapshot natively on Android without a problem. I was hoping to get confirmation from TBC guys, who are preparing their own Android SDK, but no word from them for the last month on this pull, so might as well move ahead.

@parkera, just need someone to review and get this in, then I'd like to get this into release/6.0 next. Should be perfectly safe as this only affects Android, touches no code on any other platform.

parkera commented:
@swift-ci test

jmschonfeld commented on Tests/FoundationEssentialsTests/DataIOTests.swift:
```diff
@@ -238,7 +238,7 @@ class DataIOTests : XCTestCase {
     
     func test_zeroSizeFile() throws {
-        #if !os(Linux)
+        #if !os(Linux) && !os(Android)
```
We might have discussed this before, but is there any semantic difference between the `os(Android)` used here in the test files, and the `canImport(Android)` used in the implementation?

finagolfin replied:
None, I simply followed the convention I've seen elsewhere in the toolchain source of using `canImport(randomOS)` when invoking platform-specific APIs and `os(randomOS)` when separating actual OS differences. Since this `/proc` behavior is in the OS, I used `os(Android)` here. I did not go changing conditional compilation checks for other non-Android OS's that may be doing this wrong to use the same convention though.

jmschonfeld replied:
Ah ok, sounds good - as long as we have a reasoning for it and they're semantically equivalent, this seems fine to me

jmschonfeld commented on Tests/FoundationEssentialsTests/FileManager/FileManagerTests.swift:
```diff
@@ -331,6 +335,7 @@ final class FileManagerTests : XCTestCase {
     }
     
     func testLinkFileAtPathToPath() throws {
+        #if !os(Android) // Hard links are not normally allowed on Android.
```
I don't think I see anything in the implementation that's specific to android with this function (unless that was in a previous PR). Are we missing an Android check in the implementation to throw? I think rather than avoiding the test if that's the case maybe it makes sense to `XCTAssertThrowsError` here?

jmschonfeld replied:
If calling `link()` doesn't work on Android, I think we should replace that call in the implementation to something that will work on Android, or if Android does not have that concept then we should probably throw to indicate that it did not succeed. That way, either the test will pass as-is or we can update the test to check for the thrown error on Android

finagolfin replied:
[Android doesn't allow hard linking on most partitions for regular users for the last decade](https://stackoverflow.com/questions/32365690/is-android-m-not-allowing-hard-links). Since it may work in those few remaining scenarios, I didn't really want to think about it and just disabled the test.

Since it will usually fail, I could replace it with this if you'd like instead:
```
#if os(Android)
XCTAssertEqual($0.delegateCaptures.shouldProceedAfterLinkError, [.init("foo", "bar", code: .fileWriteNoPermission)])
XCTAssertFalse($0.fileExists(atPath: "bar"))
#else
... current passing asserts ...
```

finagolfin replied:
As discussed, added those Android-specific asserts to this pull

jmschonfeld reviewed (commented):
Overall LGTM, but a few small comments

finagolfin commented on Tests/FoundationEssentialsTests/FileManager/FileManagerTests.swift:
```diff
@@ -23,6 +23,10 @@ import TestSupport
 
+#if canImport(Android)
+import Android
+#endif
```
This is the compilation error I get without it:
```
Tests/FoundationEssentialsTests/FileManager/FileManagerTests.swift:622:15: error: cannot find 'getuid' in scope
620 |     func testFileAccessAtPath() throws {
621 |         #if !os(Windows)
622 |         guard getuid() != 0 else {
    |               `- error: cannot find 'getuid' in scope
```
There are some strange issues with these platform overlays, [which I investigated a bit last month](https://github.com/apple/swift-certificates/pull/183#discussion_r1730366251). Since nobody I asked seems to have a good idea of what's going on, I chalked it up to some Swift module leak that can be fixed later and am just adding these imports now to shut the compiler up.

finagolfin commented:
Addressed all review feedback and updated, one last CI run and this can go in.

jmschonfeld commented:
@swift-ci please test

finagolfin commented:
Passed CI, ready for merge. I will submit for 6.0 next.
